### PR TITLE
chore(type-utils): reuse newly added "is builtin symbol like" logic

### LINKED ISSUE #8234: Repo: reuse newly added "is builtin symbol like" logic from `type-utils`
### Suggestion

Followup for https://github.com/typescript-eslint/typescript-eslint/pull/8011#discussion_r1417737352

We've added few utility functions to the `packages/type-utils/src/builtinSymbolLikes.ts`

Basically they contain the logic for determining whether a given type is a type from the TS default library or not

These files already contain similar logic, it'd be cool if we dedupe it!

https://github.com/typescript-eslint/typescript-eslint/blob/705370ac0d9c54081657b8855b398e57d6ea4ddb/packages/eslint-plugin/src/rules/no-implied-eval.ts#L82-L90 https://github.com/typescript-eslint/typescript-eslint/blob/705370ac0d9c54081657b8855b398e57d6ea4ddb/packages/eslint-plugin/src/rules/no-implied-eval.ts#L145-L153

Let's label this issue as `good first issue`? Ref: https://github.com/typescript-eslint/typescript-eslint/pull/8011#discussion_r1438375603

### PR DESCRIPTION
**Pull request by arka1002**
fixes: #8234

## PR Checklist

- [x] Addresses an existing open issue: fixes #8234 
- [x] That issue was marked as [accepting prs](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aopen+is%3Aissue+label%3A%22accepting+prs%22)
- [x] Steps in [Contributing](https://typescript-eslint.io/contributing) were taken

### CODE REVIEW

File: packages/type-utils/src/builtinSymbolLikes.ts
Comment (auvred) [inline packages/type-utils/src/builtinSymbolLikes.ts:78]: `typescript` package already exports the function with same name `isFunctionLike`. Having two functions with the same name may cause confusion.

File: packages/type-utils/src/builtinSymbolLikes.ts
Comment (auvred) [inline packages/type-utils/src/builtinSymbolLikes.ts:81]: The original rule didn't check for symbols with name `FunctionConstructor`. Should we add tests that contain symbols with name `FunctionConstructor`? But that is probably not within the scope of this pr. IMO we shouldn't change the rule logic in this pr

File: packages/eslint-plugin/src/rules/no-implied-eval.ts
Comment (auvred) [inline packages/eslint-plugin/src/rules/no-implied-eval.ts:85]: `isFunctionLike` internally already checks that the type has a symbol, and that symbol is named `Function`, so the condition above is redundant.

File: packages/eslint-plugin/src/rules/no-implied-eval.ts
Comment (auvred) [inline packages/eslint-plugin/src/rules/no-implied-eval.ts:1]: `@typescript-eslint/type-utils` shouldn't be imported directly. This function should probably be imported from `src/utils`, as other rules do

https://github.com/typescript-eslint/typescript-eslint/blob/920f909debc3eb394f5aef4e36568bdf6641e304/packages/eslint-plugin/src/util/index.ts#L18-L19

**Review by auvred — CHANGES_REQUESTED** [review]

File: packages/type-utils/src/builtinSymbolLikes.ts
Comment (auvred) [reply packages/type-utils/src/builtinSymbolLikes.ts:81]: You're right `Function` here has `FunctionConstructor` type. I missed this because in the source code of rule both symbol name and callee identifier's name are compared to `FUNCTION_CONSTRUCTOR`

Then +1 on renaming `isFunctionLike`... Not sure what's the best name for it. Maybe `isFunctionTypeLike` or something like this?

Also we should consider that #8094 most likely will introduce changes to `isBuiltinSymbolLike`.
As per https://github.com/typescript-eslint/typescript-eslint/pull/8094#discussion_r1449120581:
`isBuiltinSymbolLike(program, type, 'Function') || isBuiltinSymbolLike(program, type, 'FunctionConstructor')` recursively visits all subtypes twice, we can optimize to not do the same job twice. #8094 does the following:

```ts
isBuiltinSymbolLike(
  services.program,
  calleeType,
  name => name === 'PromiseConstructor' || name === 'Promise'
)
```

Perhaps we can wait until #8094 is merged and then do the same

Comment (arka1002) [reply packages/type-utils/src/builtinSymbolLikes.ts:81]: Yes, sure.

Comment (JoshuaKGoldberg) [reply packages/type-utils/src/builtinSymbolLikes.ts:81]: > But that is probably not within the scope of this pr. IMO we shouldn't change the rule logic in this pr

Yeah this is just a refactor (`chore`). If you want to change how it works that'd be a separate issue.

👍 I'll mark this as blocked on #8094. Thanks :)

Comment (JoshuaKGoldberg) [reply packages/type-utils/src/builtinSymbolLikes.ts:81]: #8094 is closed - so we can now pretend it doesn't exist. cc @arka1002, I'm un-blocking this PR. Sorry for the long wait!

**Comment by JoshuaKGoldberg** [comment]
👋 ping @arka1002, just checking in - is this PR still something you have the time and energy for?

**Comment by arka1002** [comment]
Closing it 👍

**Pull request closed without merging**
